### PR TITLE
[fc][test] FC warmup part 2: Warm up H2 conns to all server instances during metadata update

### PR DESCRIPTION
**Pull request by m-nagarajan**
## Summary
* After the initial metadata refresh, noticed spikes in latency at the very beginning both in prod and cert as well as leading to potential flaky integration tests. To make the warmup more complete, adding a method to warmup the H2 conns to all the server instances. As we are not trying to validate the metadata here but just to warmup conns beforehand, warmup is on best effort basis for now. To improve the long tail and potentially further improve the client startup speed by not waiting till the timeout for all the conns to be warmed up, more fine-grained logic wrt partitions, replicas or warmup success rate can be added which might succeed faster than waiting for all conns.
* Marked `testFastClientWithLongTailRetry` as flaky as the retry mechanism in integration tests are non-deterministic leading to flaky tests
* Added a config `metadataConnWarmupTimeoutInSeconds` defaulting to 20 seconds. This dictates how long to wait for the h2 connection warmup.

## How was this PR tested?
GH CI (added new unit tests)

## Does this PR introduce any user-facing changes?
- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (lluwm) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:238]: Just for my knowledge, what exactly is warmed up here after doing a `get` request?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (lluwm) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:235]: nit, the function returns the primitive type `boolean`, `return true;` seems natural to avoid the type conversion `Boolean` -> `boolean`, but maybe you have other preference and I think that's fine as well.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (lluwm) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:226]: If it is on the critical path, is it possible to avoid copying the set for performance?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:248]: @gaojieliu rather than using the METADATA endpoint leading to metadata fetched unnecessarily for every warmup request, using HEALTH endpoint here.

**Review by lluwm — APPROVED** [review]
Thanks for addressing the comments and LGTM!

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (gaojieliu) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:248]: SG

**Review by gaojieliu — COMMENTED** [review]
Looks good overall, left some minor comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:315]: Good point. Latest change track futures as well via `warmUpInstancesFutures` and added better logging.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:279]: Done, thanks

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java]: using `Executors.newCachedThreadPool(new NamedThreadFactory("h2ConnWarmupForFastClient"))` in the latest diff.

**Review by m-nagarajan — COMMENTED** [review]
Thanks @gaojieliu for the review. Addressed the comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:245]: Nit:
Why don't we remove it directly from the set?
```
entryIterator.remove();
```

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (gaojieliu) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:271]: The logic here is a little tricky to read..
So shall we just maintain one map: `warmedUpIntanceFutures` and keep the instance entries even the future is done?
Right now, this function is maintaining two fields: ` warmedUpInstances` and `warmUpInstancesFutures`.

**Review by gaojieliu — COMMENTED** [review]
Looks good overall, and left some minor comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (xunyin8) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:326]: NIT: maybe WARN instead of ERROR level here?

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (xunyin8) [inline clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java:315]: I see that we are doing quite a bit of logging here and there regarding to the progress of warmup. Can we standardize the logging to make sure every log contains the necessary info for us to make sense of them? e.g. we are logging about store and version for warmup start but then the error and completion logs no longer have these info.
```
        "Starting H2 connection warm up from fast client to {} instances for store {} version {}",
```

**Review by xunyin8 — COMMENTED** [review]
Sorry for the late review. Change LGTM just some minor logging comments.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java]: cleaned it up using only `warmedUpIntanceFutures`

**Review by xunyin8 — APPROVED** [review]
LGTM thanks! Let's see if Gaojie still have any comments before merging.

File: clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java
Comment (m-nagarajan) [reply clients/venice-client/src/main/java/com/linkedin/venice/fastclient/meta/RequestBasedMetadata.java]: Added a common prefix for all logs.

**Review by m-nagarajan — COMMENTED** [review]
Thanks @xunyin8 for the review. addressed the comments.

**Review by gaojieliu — APPROVED** [review]
Thanks!

**Pull request merged**
